### PR TITLE
feat: [CDS-56738]: completed the K8s BG Stage Scale down step

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
This PR contains the complete delegate side changes for K8s Blue Green Stage Scale Down Step. This will Scale down the workload present in stage environment. This step does not have any outcome and using release name as input to process the release history and bring down stage environment.

Delegate PR: [https://github.com/harness/harness-core/pull/47835](https://github.com/harness/harness-core/pull/47835)
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47989)

### CODE REVIEW

**Review by abhinavsingh1196 — APPROVED** [review]

**Review by namantalaycha10 — APPROVED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStep.java [hunk 1]
```diff
@@ -45,15 +65,42 @@ public void validateResources(Ambiance ambiance, StepElementParameters stepParam
...
+    K8sBlueGreenStageScaleDownRequest request =
+        K8sBlueGreenStageScaleDownRequest.builder()
+            .commandName(K8S_BG_STAGE_SCALE_COMMAND_NAME)
+            .releaseName(releaseName)
+            .taskType(K8sTaskType.BLUE_GREEN_STAGE_SCALE_DOWN)
+            .timeoutIntervalInMin(
+                NGTimeConversionHelper.convertTimeStringToMinutes(stepElementParameters.getTimeout().getValue()))
+            .k8sInfraDelegateConfig(cdStepHelper.getK8sInfraDelegateConfig(infrastructure, ambiance))
+            .useNewKubectlVersion(cdStepHelper.isUseNewKubectlVersion(accountId))
+            .useK8sApiForSteadyStateCheck(cdStepHelper.shouldUseK8sApiForSteadyStateCheck(accountId))
```
Comment (abosii-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStep.java:81]: Do we need this FF? We should not do steady state check

**Review by abosii-harness — CHANGES_REQUESTED** [review]
If we're good with scale down, should we mark tasktype K8S_BLUE_GREEN_STAGE_SCALE_DOWN_TASK as supported?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStep.java (hunk 1 above)
Comment (GoelPratyush99) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStep.java]: Removed it. We don't require this FF

**Review by GoelPratyush99 — COMMENTED** [review]
sure merge it to this PR

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47989#-:-NViCecd90CKjLcf1O5o:b-w4wgmy)* status: 6 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained.

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by abosii-harness — APPROVED** [review]

**Comment by GoelPratyush99** [comment]
trigger uts

File: 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStep.java [hunk 2]
```diff
@@ -45,15 +65,41 @@ public void validateResources(Ambiance ambiance, StepElementParameters stepParam
...
-    return TaskRequest.newBuilder().build();
+    InfrastructureOutcome infrastructure = (InfrastructureOutcome) outcomeService.resolve(
+        ambiance, RefObjectUtils.getOutcomeRefObject(OutcomeExpressionConstants.INFRASTRUCTURE_OUTCOME));
+    String releaseName = cdStepHelper.getReleaseName(ambiance, infrastructure);
+    String accountId = AmbianceUtils.getAccountId(ambiance);
+    K8sBlueGreenStageScaleDownRequest request =
+        K8sBlueGreenStageScaleDownRequest.builder()
+            .commandName(K8S_BG_STAGE_SCALE_COMMAND_NAME)
+            .releaseName(releaseName)
+            .taskType(K8sTaskType.BLUE_GREEN_STAGE_SCALE_DOWN)
```
Comment (TarunUba1) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStep.java:76]: nit: we can have same naming convention for this feature, K8S_BG_STAGE_SCALE_COMMAND_NAME and  BLUE_GREEN_STAGE_SCALE_DOWN feels like two things for same.

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 3 of 5 files at r1, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47989)* status: 6 file(s) to review. 1. unresolved discussions.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStep.java (hunk 2 above)
Comment (GoelPratyush99) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/K8sBGStageScaleDownStep.java:76]: I think that it's understandable. We have been following similar naming convention for all the steps "Apply", "Blue Green" etc.

**Comment by GoelPratyush99** [comment]
trigger review

**Comment by GoelPratyush99** [comment]
trigger messagecheck

**Pull request merged**